src/client/keycode.cpp

```diff
@@ -221,122 +223,177 @@ static const struct table_key table[] = {
 	DEFINEKEY5("_")
 };
 
+static const table_key invalid_key = {"", irr::KEY_UNKNOWN, L'\0', ""};
+
 #undef N_
 
 
-static const table_key &lookup_keyname(const char *name)
+static const table_key &lookup_keychar(wchar_t Char)
 {
+	if (Char == L'\0')
+		return invalid_key;
+
 	for (const auto &table_key : table) {
-		if (strcmp(table_key.Name, name) == 0)
+		if (table_key.Char == Char)
 			return table_key;
 	}
 
-	throw UnknownKeycode(name);
+	auto newsym = wide_to_utf8(std::wstring_view(&Char, 1));
+	table_key new_key {newsym, irr::KEY_KEY_CODES_COUNT, Char, newsym};
+	return table.emplace_back(std::move(new_key));
 }
 
 static const table_key &lookup_keykey(irr::EKEY_CODE key)
 {
+	if (!Keycode::isValid(key))
+		return invalid_key;
+
 	for (const auto &table_key : table) {
 		if (table_key.Key == key)
 			return table_key;
 	}
 
-	std::ostringstream os;
-	os << "<Keycode " << (int) key << ">";
-	throw UnknownKeycode(os.str().c_str());
+	return invalid_key;
 }
 
-static const table_key &lookup_keychar(wchar_t Char)
+static const table_key &lookup_keyname(std::string_view name)
 {
+	if (name.empty())
+		return invalid_key;
+
 	for (const auto &table_key : table) {
-		if (table_key.Char == Char)
+		if (table_key.Name == name)
 			return table_key;
 	}
 
-	std::ostringstream os;
-	os << "<Char " << hex_encode((char*) &Char, sizeof(wchar_t)) << ">";
-	throw UnknownKeycode(os.str().c_str());
+	auto wname = utf8_to_wide(name);
+	if (wname.empty())
+		return invalid_key;
+	return lookup_keychar(wname[0]);
 }
 
-KeyPress::KeyPress(const char *name)
+#if USE_SDL2
```

Comment:
same here. `IrrlichtDevice` has a fallback implementation so no need to make it more complicated than necessary